### PR TITLE
Users can add an item, mark it complete or remove.

### PR DESCRIPTION
**Pull request by tahakhan**
Users can add an item, mark it complete and remove.

### CODE REVIEW

File: pages/index.tsx [hunk 1]
```diff
@@ -10,23 +10,109 @@ const Home: NextPage = () => {
...
+    setIsError(false);
+
     fetch("/api/items")
     .then(response => response.json())
     .then(data => 
       {
-        setTodoList(data.result)
         setIsLoading(false)
+        if(data.result){  
+          setTodoList(data.result)
```
Comment (ovaistariq) [inline pages/index.tsx:30]: Will this also return true when am empty result set is returned

File: pages/index.tsx [hunk 2]
```diff
@@ -10,23 +10,109 @@ const Home: NextPage = () => {
...
+
+    const result: RegExpMatchArray | null = querySearch.match('^\\S.*');
+    if(result == null){
+      setWiggleError(true);
+      return;
+    }
+
+    setIsLoading(true);
+
+    fetch("/api/items", {
```
Comment (ovaistariq) [inline pages/index.tsx:62]: Wouldn't this be inserting data? Or am I misreading the code?

File: pages/api/item/[id].ts [hunk 3]
```diff
@@ -56,7 +56,7 @@ async function handlePut (
   res: NextApiResponse<Data>
 ) {
   try {
-    const item = req.body as TodoItem
+    const item = JSON.parse(req.body) as TodoItem
```
Comment (adilansari) [inline pages/api/item/[id].ts:59]: +1

File: pages/index.tsx [hunk 4]
```diff
@@ -10,23 +10,109 @@ const Home: NextPage = () => {
...
     .then(response => response.json())
     .then(data => 
       {
-        setTodoList(data.result)
         setIsLoading(false)
+        if(data.result){  
+          setTodoList(data.result)
+        }
+        else {
+          setIsError(true)
```
Comment (adilansari) [inline pages/index.tsx:33]: would this set error when response is 200 and data.result is empty?

File: pages/index.tsx [hunk 5]
```diff
@@ -10,23 +10,109 @@ const Home: NextPage = () => {
...
+    setIsLoading(true);
+  
+    fetch("/api/item/"+item.id, {
+      method:'PUT',
+      body:JSON.stringify(item)
+    })
+    .then(() => 
+      {
+        setIsLoading(false);
+        fetchListItems();
```
Comment (adilansari) [inline pages/index.tsx:108]: Should we try to leverage something like `getStaticProps` to refresh the list items? We can change the app backend likewise. I have seen patterns in next where you can directly query the database from browser. We can eliminate /api/items if we want to leverage one of those options.

Maybe worth [trying out](https://nextjs.org/docs/basic-features/data-fetching/get-static-props#write-server-side-code-directly)

Comment (ovaistariq) [reply pages/index.tsx:108]: I think it is best to let it flow through the /api endpoints. Then all the functionality is being provided through serverless functions consistently. WDYT?

Comment (adilansari) [reply pages/index.tsx:108]: Let's keep it on `/api` for now, its not anything anti-pattern. We demonstrate the code and let devs decide whichever route works best for their case. Eventually its the same function, they decide which call is optimal for them.

File: pages/index.tsx (hunk 1 above)
Comment (tahakhan) [reply pages/index.tsx:35]: Yes, the result key is always present in the response. When empty, data.results is set to an empty array [].

File: pages/index.tsx (hunk 4 above)
Comment (tahakhan) [reply pages/index.tsx:38]: No, this wouldn't.

File: pages/index.tsx (hunk 5 above)
Comment (tahakhan) [reply pages/index.tsx]: If we use getStaticProps it generates HTML/JSON at build time and passes them as props(todo list, in our case) to the component. We would be needing to refresh the list as user is interacting(adding, deleting..).

[getstaticprops](https://nextjs.org/docs/basic-features/data-fetching/get-static-props#when-should-i-use-getstaticprops)

getStaticProps runs at build time and pre-renders the props before the user lands on the page, it also has a 'revalidate()' method which looks like it runs in the background. Not sure if we can use it to refresh our list instantly when a user fetches an updated list or adds/deletes an item.

[revalidate](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation)

I can look into it further after the conference, but api/endpoints looks pretty standard.

**Review by ovaistariq — APPROVED** [review]

**Review by adilansari — APPROVED** [review]

**Pull request merged**